Add esy

jchavarri opened the pull request:
Proposal to add `esy` for a simpler dependencies management workflow.

One potential downside is that `esy` places the binaries in `./_build/default`, I'm not sure if this is an issue.

I adapted the docs a bit as well :)

jchavarri committed:
Add esy

jchavarri committed:
Update executable path

pbiggar commented on README.md:
```diff
@@ -4,19 +4,37 @@ Philip2 ("*fee-leap* the second") is a compiler from Elm to OCaml, using the Rea
+
+    % esy ./_build/default/src/philip2.exe
+
+## Build elm-format
```
This isn't quite right. This makes it seem like these instructions will build elm-format, but they're actually to translate your app. MAybe rename install-dependencies to install-elm-format, then add a new subsection called `quickstart` or something.

pbiggar reviewed (changes requested):
This is really cool, thanks! Do you have time to add the build commands to the circleci/config.yml file? If not, I can do it.

jchavarri committed:
Add circle, use esy to manage tests, update readme

jchavarri commented:
I have updated the branch with a few things: circle config (based on facebook/reason), updated readme and also updated dune and esy files so the tests can be run also from esy.

I'm running though in a problem with the tests. Everything seems to work (`.elm` and `.ml.out` files are copied in the build folder) but if you call:

```
esy install
esy build
esy test
```

I get:

```
        test alias tests/runtest (exit 1)
(cd /Users/javi/Development/github/philip2/_esy/default/store/b/philip2-0.0.1-57cc45f6/default/tests && ./test.exe)
Testing suite.
This run has ID `18A46BAA-4D18-4E7B-A7E1-FCE767AC9524`.
[ERROR]             tests          0   simple.
-- tests.000 [simple.] Failed --
in /Users/javi/Development/github/philip2/_esy/default/store/b/philip2-0.0.1-57cc45f6/default/tests/_build/_tests/18A46BAA-4D18-4E7B-A7E1-FCE767AC9524/tests.000.output:
[exception] Command failed: ../src/philip2.exe ../tests/simple.elm Exit status: exited with code 255 stderr: Raised at file "common.ml", line 5, characters 19-39
Called from file "src/translate.ml", line 965, characters 2-40
Called from file "src/philip2.ml", line 11, characters 11-51
("Yojson.Json_error(\"Blank input data\")")

Raised at file "src/shell.ml", line 176, characters 10-232
Called from file "tests/test.ml", line 20, characters 6-89
Called from file "alcotest/junit_alcotest.ml", line 10, characters 6-13
Re-raised at file "alcotest/junit_alcotest.ml", line 37, characters 6-15
Called from file "src/alcotest.ml", line 276, characters 8-14
```

It seems that somehow the `philip2.exe` binary is not expecting the `.elm` argument being passed the way I'm doing it. Could you check if everything's ok? I think I'm treating `philip2.exe` as `translate.exe` but maybe there are differences.

pbiggar commented:
The file `translate` is different from `translate.exe`. It's this file: https://github.com/darklang/philip2/blob/master/translate. The error is happening because it expects json input, not an elm file.

jchavarri committed:
Pass tests

jchavarri committed:
Fix CI

jchavarri committed:
Fix esy path in CI

jchavarri commented:
Cool, that makes sense.

With that info, I got the tests working by removing the elm-to-json step, and use `philip2` instead of `translate`. This gets rid of the dependency on `elm-format` and the whole Haskell `stack`, which makes the CI setup a little bit easier. This means the input file for the tests is already in a json format. I slightly changed the philip2 executable to also accept a file path.

One strange thing is that I'm getting some casting to `string` in the output, see the `.out` file:  https://github.com/darklang/philip2/pull/1/files#diff-44e707cb9d78aa44bfe1000dfd8c7bdaR3 (Edit: ooo i guess that's due to the `ocamlformat` pass that was done in `translate` and is not done in `philip2.exe` :) 

I had also:
- updated all the references to `./build` with `esy build`
- updated the `ocamlformat` call in `translate` to use the `esy` version, for users like in my case that don't have global `opam` packages installed 😄 
- made sure CI passes by using the `node` Docker image instead of the raw Debian one.

pbiggar commented:
Thanks for the hard work!

> With that info, I got the tests working by removing the elm-to-json step, and use philip2 instead of translate. This gets rid of the dependency on elm-format and the whole Haskell stack, which makes the CI setup a little bit easier. This means the input file for the tests is already in a json format. I slightly changed the philip2 executable to also accept a file path.

I would prefer not to do this change. For testing, we want to include the source file and not include the generated file. We also want to test to ensure that our build instructions for elm-format (which is an essential part of the tool) work. So if you don't mind would we be able to go back to using the translate script?

jchavarri committed:
Remove json-based tests

jchavarri commented:
I understand. My reasoning was that, considering `elm-format` is a different tool, it made sense to actually test what philip2 does, which is producing an `ml` file from the `json` Elm syntax tree. If the whole pipeline (elm to json to ml) needs to be tested, I would consider making the `darklang/elm-format` fork part of this same repo to keep them in sync more easily over time. I suspect that that would make testing easier too :)

I got the tests working again locally without the json (see latest commit). I had to create a copy of `translate` inside the `tests` folder, just because `esy` sandboxes all the compiled artifacts inside a folder, so keeping `translate` in the root folder made things quite hard from a path perspective (`test.exe` calls `translate`, which calls the `philip2.exe` binary). I also had to change the permissions from OCaml. Feel free to remove one of both copies or rename the new one to something else.

I don't have any more time this week to look at fixing CI to include `elm-format`. I think that just calling `./install-elm-format` before running `esy test` should work...

Looking forward to the publication of this tool! 🚀

pbiggar commented:
>  I would consider making the darklang/elm-format fork part of this same repo to keep them in sync more easily over time. I suspect that that would make testing easier too :)

Good idea.

I'll try to get this over the line this week. Thanks a million for the help.

pbiggar committed:
Add elm-format in CI

pbiggar committed:
Fix CI build

pbiggar committed:
Fix stack cache

pbiggar committed:
Don't cache off circleci config

pbiggar committed:
Fix CircleCI path

pbiggar committed:
Move run key to an actual valid place in the config file

pbiggar committed:
Make stack work

pbiggar committed:
Fix syntax

pbiggar committed:
Add env to debug test step

pbiggar committed:
Always rerun tests

pbiggar committed:
set up environment to run .bashrc

Remove circleci verbosity

pbiggar committed:
Fix errors

pbiggar committed:
Simplify caching, actually cache elm-format

pbiggar committed:
Pin esy

pbiggar committed:
Cache .stack again

pbiggar committed:
Don't install stack if it's already installed

pbiggar committed:
Re-add dependencies to make tests work

pbiggar commented:
Finally got this working.
